osu.Game.Rulesets.Taiko/UI/DrumTouchInputArea.cs

```diff
@@ -0,0 +1,141 @@
+using osu.Framework.Logging;
+using osu.Game.Graphics;
+using osu.Game.Rulesets.UI;
+using osuTK;
+
+namespace osu.Game.Rulesets.Taiko.UI
+{
+    /// <summary>
+    /// An overlay that captures and displays Taiko mouse and touch input.
+    /// The boundaries of this overlay defines the interactable area for touch input.
+    /// A secondary InputDrum is attached by this overlay, which defines the circulary boundary which distinguishes "centre" from "rim" hits, and also displays input.
+    /// </summary>
+    public class DrumTouchInputArea : Container
+    {
+        // The percent of the drum that extends past the bottom of the screen (set to 0.0f to show the full drum)
+        private const float overhangPercent = 0.33f;
+        private readonly InputDrum touchInputDrum;
+
+        [Resolved(canBeNull: true)]
+        private TaikoInputManager taikoInputManager { get; set; }
+
+        private KeyBindingContainer<TaikoAction> keyBindingContainer;
+
+        // Which Taiko action was pressed by the last OnMouseDown event, so that the corresponding action can be released OnMouseUp even if the cursor position moved
+        private TaikoAction mouseAction;
+
+        // A map of (Finger Index OnTouchDown -> Which Taiko action was pressed), so that the corresponding action can be released OnTouchUp is released even if the touch position moved
+        private Dictionary<TouchSource, TaikoAction> touchActions = new Dictionary<TouchSource, TaikoAction>(Enum.GetNames(typeof(TouchSource)).Length);
+
+        private Playfield playfield;
+
+        public DrumTouchInputArea(Playfield playfield) {
+            this.playfield = playfield;
+
+            RelativeSizeAxes = Axes.Both;
+            RelativePositionAxes = Axes.Both;
+            Children = new Drawable[]
+            {
+                new Box() {
+                    Alpha = 0.0f,
+                    Colour = new OsuColour().Blue,
+
+                    RelativeSizeAxes = Axes.Both,
+                    RelativePositionAxes = Axes.Both,
+                },
+                new Container() {
+                    RelativeSizeAxes = Axes.Both,
+                    RelativePositionAxes = Axes.Both,
+
+                    Anchor = Anchor.BottomCentre,
+                    Origin = Anchor.BottomCentre,
+
+                    Children = new Drawable[]
+                    {
+                        touchInputDrum = new InputDrum(playfield.HitObjectContainer) {
+                            Anchor = Anchor.Centre,
+                            Origin = Anchor.Centre,
+                        },
+                    }
+                },
+            };
+
+        }
+        protected override void LoadComplete()
+        {
+            keyBindingContainer = taikoInputManager?.KeyBindingContainer;
+
+            Padding = new MarginPadding {
+                Top =  playfield.ScreenSpaceDrawQuad.BottomLeft.Y,
+                Bottom = -DrawHeight * overhangPercent,
+            };
```

Comment:
I have no idea what this means, but I don't think it can exist like this.

Reply:
Sure, yeah I'll clean this up.

Essentially the Top padding is there so that the visual elements start right below the playfield.
The negative bottom padding is so that the drum is shown going a bit off the bottom of the screen, similar to how it's shown in Taiko no Tatsujin games (as opposed to the full drum being in view)

Reply:
Changed as of 317869078f470889330702b7b99b6146569b4f75.